File: MiniDB_Projects/Team_Alpha/src/minidb/txn/LockManager.java
package minidb.txn;

import java.util.*;

/**
 * LockManager implements Two-Phase Locking with two lock modes:
 *
 *   SHARED (S)    — multiple readers may hold it simultaneously.
 *   EXCLUSIVE (X) — only one writer; incompatible with any other lock.
 *
 * Compatibility matrix:
 *            | S        | X
 *       -----+----------+--------
 *        S   | grant    | block
 *        X   | block    | block
 *
 * Locks are taken on a "resource" identified by a string key (here: "table:RID"
 * for row-level locking, or "table" for coarse table locks).
 *
 * DEADLOCK DETECTION: we maintain a wait-for graph (txn -> txns it waits on).
 * Before a transaction blocks, we check whether granting the wait would create
 * a cycle. If so, we abort the requesting transaction (the "victim") and throw
 * DeadlockException, which the executor turns into a rollback.
 */
public final class LockManager {
    public enum Mode { SHARED, EXCLUSIVE }

    public static final class DeadlockException extends RuntimeException {
        public DeadlockException(String m) { super(m); }
    }

    private static final class LockEntry {
        final Map<Long, Mode> holders = new HashMap<>();   // txnId -> mode
        final List<long[]> waitQueue = new ArrayList<>();  // [txnId, modeOrdinal]
    }

    private final Map<String, LockEntry> table = new HashMap<>();
    // wait-for graph: who is each txn waiting on
    private final Map<Long, Set<Long>> waitsFor = new HashMap<>();

    public synchronized void acquire(Transaction txn, String resource, Mode mode) {
        // lock upgrade / re-entrancy: already hold something on this resource?
        Mode held = txn.heldLocks.get(resource);
        if (held == Mode.EXCLUSIVE) return;
        if (held == Mode.SHARED && mode == Mode.SHARED) return;

        LockEntry e = table.computeIfAbsent(resource, k -> new LockEntry());

        while (!compatible(e, txn.id, mode)) {
            // record who we are waiting for, then check for a deadlock cycle
            Set<Long> blockers = new HashSet<>(e.holders.keySet());
            blockers.remove(txn.id);
            waitsFor.put(txn.id, blockers);
            if (hasCycle(txn.id)) {
                waitsFor.remove(txn.id);
                throw new DeadlockException("Deadlock detected; aborting " + txn);
            }
            try { wait(200); } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(ie);
            }
        }
        waitsFor.remove(txn.id);
        e.holders.put(txn.id, mode);
        txn.heldLocks.put(resource, mode);
    }

    private boolean compatible(LockEntry e, long txnId, Mode mode) {
        for (Map.Entry<Long, Mode> h : e.holders.entrySet()) {
            if (h.getKey() == txnId) continue; // ignore self
            if (mode == Mode.EXCLUSIVE || h.getValue() == Mode.EXCLUSIVE) return false;
        }
        return true;
    }

    /** Release every lock held by a transaction (called at commit/abort). */
    public synchronized void releaseAll(Transaction txn) {
        for (String res : txn.heldLocks.keySet()) {
            LockEntry e = table.get(res);
            if (e != null) e.holders.remove(txn.id);
        }
        txn.heldLocks.clear();
        waitsFor.remove(txn.id);
        notifyAll(); // wake any waiters
    }

    // DFS cycle detection in the wait-for graph starting from `start`.
    private boolean hasCycle(long start) {
        Set<Long> visited = new HashSet<>();
        Deque<Long> stack = new ArrayDeque<>();
        stack.push(start);
        boolean first = true;
        while (!stack.isEmpty()) {
            long cur = stack.pop();
            if (!first && cur == start) return true;
            first = false;
            if (!visited.add(cur)) continue;
            Set<Long> next = waitsFor.get(cur);
            if (next != null) for (long n : next) stack.push(n);
        }
        return false;
    }
}

File: MiniDB_Projects/Team_SuperSymmetry/src/minidb/lock_manager.py
"""
Thread-safe lock manager for Strict Two-Phase Locking (Strict 2PL).

Row-granularity shared (S) / exclusive (X) locks, identified by an opaque
resource id (e.g. (table, rid)). Locks are held until commit/abort (strict
2PL) -> conflict-serializable, strict (recoverable, cascadeless) schedules.

This is a genuinely *blocking* manager: a transaction that requests a
conflicting lock waits on a condition variable until the lock is released.
Before a transaction is allowed to wait, the manager checks the global
wait-for graph for a cycle; if waiting would deadlock, a DeadlockError is
raised and the requesting transaction becomes the victim.
"""
from __future__ import annotations

import threading
from collections import defaultdict
from enum import Enum
from typing import Dict, List, Set


class LockMode(Enum):
    S = "S"
    X = "X"


class DeadlockError(Exception):
    pass


class _Resource:
    __slots__ = ("holders", "waiters")

    def __init__(self):
        self.holders: Dict[int, LockMode] = {}   # txn -> mode
        self.waiters: List[int] = []             # txns queued (FIFO)


class LockManager:
    def __init__(self):
        self._mutex = threading.Condition()
        self._res: Dict[object, _Resource] = defaultdict(_Resource)
        self._held_by: Dict[int, Set[object]] = defaultdict(set)
        # wait-for graph: txn -> set of txns it is currently waiting on
        self._waits_for: Dict[int, Set[int]] = defaultdict(set)

    # ---- helpers (call with mutex held) ---------------------------------
    def _compatible(self, res: _Resource, txn_id: int, mode: LockMode) -> bool:
        for other, omode in res.holders.items():
            if other == txn_id:
                continue
            if mode == LockMode.X or omode == LockMode.X:
                return False
        return True

    def _creates_cycle(self, txn_id: int, blockers: Set[int]) -> bool:
        # would txn_id -> blockers introduce a cycle reaching txn_id?
        seen: Set[int] = set()
        stack = list(blockers - {txn_id})
        while stack:
            t = stack.pop()
            if t == txn_id:
                return True
            if t in seen:
                continue
            seen.add(t)
            stack.extend(self._waits_for.get(t, ()))
        return False

    # ---- public ----------------------------------------------------------
    def acquire(self, txn_id: int, resource, mode: LockMode, timeout: float = 5.0):
        with self._mutex:
            res = self._res[resource]
            # upgrade or re-entrant
            if txn_id in res.holders:
                if res.holders[txn_id] == LockMode.X or mode == LockMode.S:
                    return
                # S -> X upgrade
                while not self._compatible(res, txn_id, LockMode.X):
                    blockers = {h for h in res.holders if h != txn_id}
                    if self._creates_cycle(txn_id, blockers):
                        raise DeadlockError(
                            f"deadlock: txn {txn_id} upgrading {resource}")
                    self._waits_for[txn_id] = blockers
                    if not self._mutex.wait(timeout):
                        self._waits_for.pop(txn_id, None)
                        raise DeadlockError(
                            f"timeout: txn {txn_id} upgrading {resource}")
                self._waits_for.pop(txn_id, None)
                res.holders[txn_id] = LockMode.X
                return

            while not self._compatible(res, txn_id, mode):
                blockers = {h for h in res.holders if h != txn_id}
                if self._creates_cycle(txn_id, blockers):
                    raise DeadlockError(
                        f"deadlock: txn {txn_id} waiting for {resource} "
                        f"held by {sorted(blockers)}")
                self._waits_for[txn_id] = blockers
                if txn_id not in res.waiters:
                    res.waiters.append(txn_id)
                if not self._mutex.wait(timeout):
                    self._waits_for.pop(txn_id, None)
                    if txn_id in res.waiters:
                        res.waiters.remove(txn_id)
                    raise DeadlockError(
                        f"timeout: txn {txn_id} waiting for {resource}")
            self._waits_for.pop(txn_id, None)
            if txn_id in res.waiters:
                res.waiters.remove(txn_id)
            res.holders[txn_id] = mode
            self._held_by[txn_id].add(resource)

    def release_all(self, txn_id: int):
        with self._mutex:
            for resource in list(self._held_by.get(txn_id, ())):
                res = self._res.get(resource)
                if res:
                    res.holders.pop(txn_id, None)
            self._held_by.pop(txn_id, None)
            self._waits_for.pop(txn_id, None)
            # this txn may have been someone's blocker -> wake everyone
            self._mutex.notify_all()
